### PR TITLE
Implemented the Guest Lifecycle (Shutdown, Start, Suspend, Resume) with device pass-through.

### PR DESCRIPTION
**Pull request by santwana**
Implemented the Guest Lifecycle (Shutdown, Start, Suspend, Resume) with device pass-through.
The guest life-cycle can be achieved for the desired iterations along with device pass-through.

Signed-off-by: Santwana Samantray <santwana@linux.vnet.ibm.com>

### CODE REVIEW

File: libvirt/tests/cfg/libvirt_pci_passthrough_lifecycle.cfg [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+            libvirt_pci_net_ip = "ENTER.YOUR.IP"
+            # Please enter a vailable ip from the net device.
+            # We need to ping it after attaching pci device
+            # to guest to verify this device works well in guest.
+            libvirt_pci_server_ip = "ENTER.YOUR.SERVER.IP"
+            libvirt_pci_net_dev_name = "ENTER.YOUR.DEVICE.NAME"
+            # Enter netmask in CIDR notation
+            libvirt_pci_net_mask = "ENTER.YOUR.NETMASK"
+        - STORAGE:
+            libvirt_pci_device_type = "STORAGE"
```
Comment (chunfuwen) [inline libvirt/tests/cfg/libvirt_pci_passthrough_lifecycle.cfg]: I don't think STORAGE can match with SRIOV, you may need add only Normal_passthrough here to filter it

File: libvirt/tests/src/libvirt_pci_passthrough_lifecycle.py [hunk 2]
```diff
@@ -0,0 +1,264 @@
+import re
+import logging
+import netaddr
+
+from virttest import virsh
+from virttest import utils_test
```
Comment (chunfuwen) [inline libvirt/tests/src/libvirt_pci_passthrough_lifecycle.py]: please try to group similar imports together.

**Comment by balamuruhans** [comment]
@santwana please squash it to single commit

File: libvirt/tests/cfg/libvirt_pci_passthrough_lifecycle.cfg [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+                    libvirt_pci_SRIOV = no
+    variants:
+        - @default:
+        - passthrough_reboot:
+            reboot_operation = "yes"
+            supported_err = "not supported by the connection driver: virDomainReboot"
+        - passthrough_suspend:
+            suspend_operation = "yes"
+        - passthrough_shutdown_start:
+            shutdown_operation = "yes"
```
[the remark this answers is not in the document]
Comment (santwana) [reply libvirt/tests/cfg/libvirt_pci_passthrough_lifecycle.cfg]: @balamuruhans Made changes, kindly review

**Comment by iccaszhulili** [comment]
rh-caselink retest please

File: libvirt/tests/src/libvirt_pci_passthrough_lifecycle.py [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+        4. Start guest and get the result of 'fdisk -l' on guest.
+        5. Compare the result of 'fdisk -l' before and after
+            attaching storage pci device to guest.
+    """
+    def vm_func_state(desired_state):
+        result = virsh.domstate(vm_name)
+        libvirt.check_exit_status(result)
+        vm_state = result.stdout.strip()
+        if vm_state == desired_state:
+            logging.info("VM state is expected: %s" % vm_state)
```
[the remark this answers is not in the document]
Comment (santwana) [reply libvirt/tests/src/libvirt_pci_passthrough_lifecycle.py]: @balamuruhans Addressed the changes suggested, please review.

File: libvirt/tests/src/libvirt_pci_passthrough_lifecycle.py [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+                pci_xml = NodedevXML.new_from_dumpxml(val)
+                pci_address = pci_xml.cap.get_address_dict()
+                vmxml.add_hostdev(pci_address)
+
+    elif device_type == "STORAGE":
+        # Store the result of "fdisk -l" in guest.
+        if not vm.is_alive():
+            vm.start()
+        session = vm.wait_for_login()
+        output = session.cmd_output("fdisk -l|grep \"Disk identifier:\"")
```
[the remark this answers is not in the document]
Comment (santwana) [reply libvirt/tests/src/libvirt_pci_passthrough_lifecycle.py]: @balamuruhans Addressed the changes suggested, please review. Thanks.

**Comment by santwana** [comment]
@balamuruhans I have addressed the review changes and also rebased the base file to existing (#2179). Kindly review. Thanks.

File: libvirt/tests/src/libvirt_pci_passthrough.py [hunk 6]
```diff
@@ -27,20 +29,94 @@ def run(test, params, env):
...
+
+    def check_nic():
+        session = vm.wait_for_login()
+        output = session.cmd_output("lspci -D")
+        nic_list_after = output.splitlines()
+        logging.info(nic_list_after)
+        return sorted(nic_list_after) != sorted(nic_list_before)
+        session.close()
+
+    def test_suspend():
```
Comment (sathnaga) [inline libvirt/tests/src/libvirt_pci_passthrough.py]: test_suspend,reboot,shutdown, can be wrapped into single method, eg:- https://github.com/autotest/tp-libvirt/blob/master/libvirt/tests/src/libvirt_vcpu_plug_unplug.py#L42

and keep it in common place, looks like you have to use it in two test files, 
why similar scenarios in two test files?, that can be merged into same test file,

Comment (santwana) [reply libvirt/tests/src/libvirt_pci_passthrough.py]: @sathnaga Addressed the changes suggested, also removed the redundant test file. Now the proposed changes are handled in the base libvirt_pci_passthrough.py.

File: libvirt/tests/src/libvirt_pci_passthrough.py [hunk 7]
```diff
@@ -27,20 +29,95 @@ def run(test, params, env):
         4. Start guest and get the result of 'fdisk -l' on guest.
         5. Compare the result of 'fdisk -l' before and after
             attaching storage pci device to guest.
+    c). Lifecycle:
+        1. Perform the below Lifecycle events along with device pass-through.
+            i) Reboot.
+            ii) Suspend/Resume.
+            iii) Start/Shutdown.
     """
+    def vm_func_state(desired_state):
```
Comment (balamuruhans) [inline libvirt/tests/src/libvirt_pci_passthrough.py]: we can reuse `check_vm_state()` in libvirt.py

File: libvirt/tests/src/libvirt_pci_passthrough.py [hunk 8]
```diff
@@ -27,20 +29,95 @@ def run(test, params, env):
...
+        result = virsh.domstate(vm_name)
+        libvirt.check_exit_status(result)
+        vm_state = result.stdout.strip()
+        if vm_state == desired_state:
+            logging.info("VM state is expected: %s", vm_state)
+        else:
+            test.fail("VM state is unexpected: %s", vm_state)
+
+    def check_nic():
```
Comment (balamuruhans) [inline libvirt/tests/src/libvirt_pci_passthrough.py]: we can reuse `vm.get_pci_devices()`

File: libvirt/tests/src/libvirt_pci_passthrough.py [hunk 9]
```diff
@@ -27,20 +29,95 @@ def run(test, params, env):
...
+                libvirt.check_exit_status(result)
+            vm.wait_for_login()
+            if check_nic():
+                logging.debug("Adapter found after reboot")
+            else:
+                test.fail("Passthroughed adapter not found after reboot")
+        else:
+            logging.debug("No life-cycle events for the guest")
+
+    def reset_vf(pci_pf):
```
Comment (balamuruhans) [inline libvirt/tests/src/libvirt_pci_passthrough.py]: It is better to call `set_vf()` directly without this method.

File: libvirt/tests/src/libvirt_pci_passthrough.py [hunk 10]
```diff
@@ -138,58 +194,67 @@ def run(test, params, env):
...
+            vm.start()
+            session = vm.wait_for_login()
+            # The Network configuration is generic irrespective of PF or SRIOV VF
+            if device_type == "NIC":
+                if check_nic():
+                    logging.debug("Adapter passthroughed to guest successfully")
+                else:
+                    test.fail("Passthrough adapter not found in guest.")
+                net_ip = netaddr.IPAddress(net_ip)
+                output = session.cmd_output("lspci -D")
```
Comment (balamuruhans) [inline libvirt/tests/src/libvirt_pci_passthrough.py]: vm.get_pci_devices()

**Comment by santwana** [comment]
@balamuruhans Implemented the changes suggested, please review. Thanks!

File: libvirt/tests/src/libvirt_pci_passthrough.py (hunk 8 above)
Comment (balamuruhans) [reply libvirt/tests/src/libvirt_pci_passthrough.py]: @santwana, this is not changed yet

File: libvirt/tests/src/libvirt_pci_passthrough.py [hunk 11]
```diff
@@ -27,20 +29,84 @@ def run(test, params, env):
...
+            iii) Start/Shutdown.
     """
+    def check_nic():
+        session = vm.wait_for_login()
+        nic_list_after = vm.get_pci_devices()
+        logging.info(nic_list_after)
+        return sorted(nic_list_after) != sorted(nic_list_before)
+        session.close()
+
+    guest_state = libvirt.MigrationTest()
```
Comment (balamuruhans) [inline libvirt/tests/src/libvirt_pci_passthrough.py]: request to place `check_vm_state()` outside `MigrationTest()` as it is helpful in generic way,
1. sendout patch to avocado-vt by moving it out of class and make change to respective avocado-vt users.
2. sendout patch to tp-libvirt, testcases that uses check_vm_state() to adopt your changes in avocado-vt.

Thanks!

Comment (santwana) [reply libvirt/tests/src/libvirt_pci_passthrough.py]: @balamuruhans I have addressed the changes suggested:

    1. sendout patch to avocado-vt by moving it out of class and make change to respective avocado-vt users. (https://github.com/avocado-framework/avocado-vt/pull/2172).
    2. sendout patch to tp-libvirt, testcases that uses check_vm_state() to adopt your changes in avocado-vt. (https://github.com/autotest/tp-libvirt/pull/2251).

Also made necessary changes to the libvirt_pci_passthrough. 

Please review. Thanks!!

**Comment by sathnaga** [comment]
@santwana needs rebase..

**Comment by santwana** [comment]
@sathnaga @balamuruhans I have rebased this PR, also made necessary changes for (https://github.com/avocado-framework/avocado-vt/pull/2172/) and (https://github.com/autotest/tp-libvirt/pull/2251). Please review. Thanks!!

File: libvirt/tests/src/libvirt_pci_passthrough.py (hunk 8 above)
Comment (balamuruhans) [reply libvirt/tests/src/libvirt_pci_passthrough.py]: This need to be addressed.

**Comment by balamuruhans** [comment]
@santwana, request for your change w.r.t -  https://github.com/autotest/tp-libvirt/pull/1623/files#r301426044

**Comment by santwana** [comment]
> @santwana, request for your change w.r.t - https://github.com/autotest/tp-libvirt/pull/1623/files#r301426044

@balamuruhans Addressed the changes, please review. Thanks!!

**Comment by santwana** [comment]
@balamuruhans Implemented the changes suggested, please review. Thanks!!
Test-Results:
**SR-IOV**
 (1/6) guest_import.qemu.vf_assignable.qcow2.virtio_scsi.smp2.virtio_net.Guest.Ubuntu.18.04.ppc64le.powerkvm-qemu.unattended_install.import.import.default_install.aio_native: PASS (70.37 s)
 (2/6) device_passthrough.qemu.vf_assignable.qcow2.virtio_scsi.smp2.virtio_net.Guest.Ubuntu.18.04.ppc64le.powerkvm-libvirt.libvirt_pci_passthrough.NIC.SRIOV: PASS (344.42 s)
 (3/6) device_passthrough.qemu.vf_assignable.qcow2.virtio_scsi.smp2.virtio_net.Guest.Ubuntu.18.04.ppc64le.powerkvm-libvirt.libvirt_pci_passthrough.passthrough_reboot.NIC.SRIOV: PASS (428.18 s)
 (4/6) device_passthrough.qemu.vf_assignable.qcow2.virtio_scsi.smp2.virtio_net.Guest.Ubuntu.18.04.ppc64le.powerkvm-libvirt.libvirt_pci_passthrough.passthrough_suspend.NIC.SRIOV: PASS (358.90 s)
 (5/6) device_passthrough.qemu.vf_assignable.qcow2.virtio_scsi.smp2.virtio_net.Guest.Ubuntu.18.04.ppc64le.powerkvm-libvirt.libvirt_pci_passthrough.passthrough_shutdown_start.NIC.SRIOV: PASS (463.15 s)
 (6/6) guest_remove.qemu.vf_assignable.qcow2.virtio_scsi.smp2.virtio_net.Guest.Ubuntu.18.04.ppc64le.powerkvm-libvirt.remove_guest.without_disk: PASS (3.98 s)

Normal Pass-through:
(1/6) guest_import.qemu.vf_assignable.qcow2.virtio_scsi.smp2.virtio_net.Guest.Ubuntu.18.04.ppc64le.powerkvm-qemu.unattended_install.import.import.default_install.aio_native: PASS (70.18 s)
 (2/6) device_passthrough.qemu.vf_assignable.qcow2.virtio_scsi.smp2.virtio_net.Guest.Ubuntu.18.04.ppc64le.powerkvm-libvirt.libvirt_pci_passthrough.NIC.Normal_passthrough: PASS (187.20 s)
 (3/6) device_passthrough.qemu.vf_assignable.qcow2.virtio_scsi.smp2.virtio_net.Guest.Ubuntu.18.04.ppc64le.powerkvm-libvirt.libvirt_pci_passthrough.passthrough_reboot.NIC.Normal_passthrough: PASS (274.53 s)
 (4/6) device_passthrough.qemu.vf_assignable.qcow2.virtio_scsi.smp2.virtio_net.Guest.Ubuntu.18.04.ppc64le.powerkvm-libvirt.libvirt_pci_passthrough.passthrough_suspend.NIC.Normal_passthrough: PASS (211.05 s)
 (5/6) device_passthrough.qemu.vf_assignable.qcow2.virtio_scsi.smp2.virtio_net.Guest.Ubuntu.18.04.ppc64le.powerkvm-libvirt.libvirt_pci_passthrough.passthrough_shutdown_start.NIC.Normal_passthrough: PASS (310.61 s)
 (6/6) guest_remove.qemu.vf_assignable.qcow2.virtio_scsi.smp2.virtio_net.Guest.Ubuntu.18.04.ppc64le.powerkvm-libvirt.remove_guest.without_disk: PASS (4.00 s)

**Comment by balamuruhans** [comment]
Let's merge the patch!

**Pull request merged**